### PR TITLE
Host query GraphQL integration

### PR DESCRIPTION
**Pull request by Glutexo**
Add [_xjoin-search_](https://github.com/RedHatInsights/xjoin-search) as an alternative data source for the host search query. This is configured by [an environment variable](https://github.com/Glutexo/insights-host-inventory/blob/bd014f0faed3bf7c0d4eef15534ef303c754b984/app/config.py#L67). Extract the _xjoin-search_ GraphQL code from the tags queries. Extract common parts from the database queries.

### CODE REVIEW

**Comment by Glutexo** [comment]
Squashed and rebased on the [current](https://github.com/RedHatInsights/insights-host-inventory/tree/45851897d8ebcc3d71870ea89867ccfc79679d1c) [_master_](https://github.com/RedHatInsights/insights-host-inventory). Tests are failing. Plan is:

1. Finish the executive part.
2. Fix the existing tests.
3. Write new tests.

File: api/query_db.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
+from datetime import datetime
```
Comment (jharting) [inline api/query_db.py]: I wonder if `query_db` is a good name. Right now we only have a single controller (host) that take care of a single entity type (host). When this gets expanded to cover multiple entities there is going to be a need for multiple query files. Therefore, I think the filename should reflect the entity (e.g. host_query_db)

File: api/query_xjoin.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    parameters = [
+        Parameter("limit", "$limit"),
+        Parameter("offset", "$offset"),
+        Parameter("order_by", "$order_by"),
+        Parameter("order_how", "$order_how"),
+    ]
+    items = (
+        simple_object("meta", ("total",)),
+        simple_object("data", ("id", "account", "display_name", "created_on", "modified_on", "canonical_facts")),
+    )
```
Comment (jharting) [inline api/query_xjoin.py]: This looks overly complicated. I don't think it is necessary to build the query object and the parameters from scratch every time. You can define the query once (with variable placeholders) and save it as a constant. Each request will then need to build a list of parameters only. These will be used together with the constant query.

Comment (Glutexo) [reply api/query_xjoin.py]: Some parts of the query still need to be dynamic. I don’t like an idea of having the query partially pre-composed and partially dynamic much. But if you insist on that, I can use Jinja instead. 🤷🏻‍♂️

Comment (jharting) [reply api/query_xjoin.py]: > Some parts of the query still need to be dynamic.
What parts need to be dynamic?

Comment (Glutexo) [reply api/query_xjoin.py]: Filtering.

Comment (jharting) [reply api/query_xjoin.py]: They don't. You can define all the variables in the static template and then only pass in the parameters that you need.

Comment (Glutexo) [reply api/query_xjoin.py]: It looks like I was assuming the GraphQL server is dumber than it actually is. [Your code](https://github.com/RedHatInsights/insights-host-inventory/pull/526/files#diff-dc4308fba8c70f6d0b9ab17c57fb8ab2R3092) shows that a need for a smart composition is not that strong as it is possible to pass a complex object as a variable. That’s great news, means you’re right and yes – it’s possible to simplify this! 🎉

Comment (jharting) [reply api/query_xjoin.py]: perfect! :tada:

**Comment by Glutexo** [comment]
Added:

- Filtering by Insights ID. It’s the last exclusive condition. Tags and staleness remain. https://github.com/RedHatInsights/insights-host-inventory/pull/532/commits/1e4c2e0c4332db838cf7f2735170e5f84b37a949
- _ansible_host_, _reporter_ and _stale_timestamp_ to the results. https://github.com/RedHatInsights/insights-host-inventory/pull/532/commits/068fa4ab354c0cf6d4a44a090946fecf28b626f5

_tags_ are not present as they are not in the cross-join service. Falsely returned as an empty array.

**Comment by jharting** [comment]
> tags are not present as they are not in the cross-join service.
Not sure what you mean. You should be able to obtain tags from xjoin-search https://github.com/RedHatInsights/xjoin-search/blob/master/src/schema/schema.graphql#L116

**Comment by Glutexo** [comment]
Working querying by tags and by staleness. The logic of translation of the _staleness_ param to the _stale_timestamp_ comparison logic is shared between the database and cross-join subsystem. The cross-join part can be shared between the host query and the tag query. Based on the code from #548.

**Comment by Glutexo** [comment]
First version with passing tests! 🎉

Also, merged with current master and that means with #548. Refactored that a little, so the logic is DRY. I am aware that some [bits](https://github.com/Glutexo/insights-host-inventory/blob/96270e6394814e530bc53290765b594d6a217ef5/app/culling.py#L67) are now rather cryptic. I preferred not to delay further work by subsequent refactoring for simplicity.

Going to add some tests for the cross-join stuff. That should be the last thing to this pull request.

**Comment by Glutexo** [comment]
There are now tests. Ready for review.

**Comment by Glutexo** [comment]
Okay, ignoring a _Push failed_ message was not that great idea. 😅 Fixed the push.

**Comment by Glutexo** [comment]
Also rebased on current master.

File: api/host.py [hunk 3]
```diff
@@ -148,105 +122,26 @@ def get_host_list(
...
+                display_name, fqdn, hostname_or_id, insights_id, tags, page, per_page, order_by, order_how, staleness
+            )
+        except ValueError as e:
+            flask.abort(400, str(e))
+    elif config.bulk_query_source == BulkQuerySource.xjoin:
+        host_list, total = get_host_list_xjoin(
+            display_name, fqdn, hostname_or_id, insights_id, tags, page, per_page, order_by, order_how, staleness
         )
+    else:
+        raise RuntimeError("Invalid bulk query source.")
```
Comment (jharting) [inline api/host.py]: This looks unnecessary - an invalid bulk query source would make the app fail at startup in config.py - no need for runtime check

File: api/host.py [hunk 4]
```diff
@@ -148,105 +122,26 @@ def get_host_list(
...
+    if config.bulk_query_source == BulkQuerySource.db:
+        try:
+            host_list, total = get_host_list_db(
+                display_name, fqdn, hostname_or_id, insights_id, tags, page, per_page, order_by, order_how, staleness
+            )
+        except ValueError as e:
+            flask.abort(400, str(e))
+    elif config.bulk_query_source == BulkQuerySource.xjoin:
+        host_list, total = get_host_list_xjoin(
+            display_name, fqdn, hostname_or_id, insights_id, tags, page, per_page, order_by, order_how, staleness
```
Comment (jharting) [inline api/host.py]: given that the function signatures are identical there's no need to duplicate function invocation

```suggestion
    get_host_list = get_host_list_db if config.bulk_query_source == BulkQuerySource.db else get_host_list_xjoin

    try:
        host_list, total = get_host_list(
            display_name, fqdn, hostname_or_id, insights_id, tags, page, per_page, order_by, order_how, staleness
        )
    except ValueError as e:
        flask.abort(400, str(e))
```

File: api/host_query_xjoin.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    $order_how: ORDER_DIR,
+    $filter: HostFilter!
+) {
+    hosts(
+        limit: $limit,
+        offset: $offset,
+        order_by: $order_by,
+        order_how: $order_how,
+        filter: {
+            OR: [
```
Comment (jharting) [inline api/host_query_xjoin.py]: why is this top level OR operation needed?

File: app/auth/__init__.py [hunk 6]
```diff
@@ -1,15 +1,13 @@
...
 
 from api.metrics import login_failure_count
-from app import IDENTITY_HEADER
 from app.auth.identity import from_auth_header
 from app.auth.identity import from_bearer_token
 from app.auth.identity import validate
 from app.logging import get_logger
 
-__all__ = ["current_identity", "bearer_token_handler", "authentication_header_handler"]
+__all__ = ("authentication_header_handler", "bearer_token_handler", "authenticated_request")
```
Comment (jharting) [inline app/auth/__init__.py]: is `"authenticated_request"` a leftover?

File: app/serialization.py [hunk 7]
```diff
@@ -48,11 +48,29 @@ def deserialize_host(raw_data):
...
-def serialize_host(host, staleness_offset):
+def deserialize_host_xjoin(data):
+    host = Host(
+        canonical_facts=data["canonical_facts"],
+        display_name=data["display_name"],
+        ansible_host=data["ansible_host"],
+        account=data["account"],
+        facts={},
+        tags={},
+        system_profile_facts={},
```
Comment (jharting) [inline app/serialization.py]: These can be fetched from xjoin-search (disregard the schema comment). https://github.com/RedHatInsights/xjoin-search/blob/master/src/schema/schema.graphql#L114

File: app/xjoin.py [hunk 8]
```diff
@@ -23,9 +25,8 @@ def graphql_query(query_string, variables):
     url_ = url()
     logger.info("QUERY: URL %s; query %s, variables %s", url_, query_string, variables)
```
Comment (jharting) [inline app/xjoin.py]: these should be debug level

File: run.py [hunk 9]
```diff
@@ -4,7 +4,7 @@
 from app import create_app
 
 config_name = os.getenv("APP_SETTINGS", "development")
-application = create_app(config_name, start_tasks=True, start_payload_tracker=True)
```
Comment (jharting) [inline run.py]: is this intentional?

File: logconfig.ini [hunk 10]
```diff
@@ -24,7 +24,7 @@ propagate=1
 qualname=gunicorn.access
 
 [logger_sqlalchemy.engine]
-level=WARNING
```
Comment (jharting) [inline logconfig.ini]: is this intentional?

File: api/tag.py [hunk 11]
```diff
@@ -48,17 +49,9 @@
...
-        filter["gte"] = gte.isoformat()
-    if lte:
-        filter["lte"] = lte.isoformat()
-    return {"stale_timestamp": filter}
-
-
 def is_enabled():
-    return flask.current_app.config["INVENTORY_CONFIG"].bulk_query_source == BulkQuerySource.xjoin
+    config = inventory_config()
+    return config == BulkQuerySource.xjoin
```
Comment (jharting) [inline api/tag.py]:
```suggestion
    return config.bulk_query_source == BulkQuerySource.xjoin
```

File: api/host.py (hunk 3 above)
Comment (Glutexo) [reply api/host.py]: This objection is valid, my personal taste however warns be before non-exhaustive if/elif/else statements. (Some programing languages, e.g. Swift, even fail to compile in such case.) I‘ll refactor a bit.

File: api/host.py (hunk 4 above)
Comment (Glutexo) [reply api/host.py]: You mean like, inline the contents of that function? I slightly disagree here. There is quite a lot of logic in both function and having them extracted improves testability. Also there is a nice symmetry between the _db_ and _xjoin_ variants.

I‘d rather leave it as it is for now, but I agree that there is a room for improvement. Let’s rather leave that for an explicit refactoring. There is more of the code shuffling and unspaghetizing that would this code utilize.

File: api/host_query_xjoin.py (hunk 5 above)
Comment (Glutexo) [reply api/host_query_xjoin.py]: Oh. Really. It doesn’t do any harm, but is redundant. Nice catch!

File: api/host.py (hunk 4 above)
Comment (jharting) [reply api/host.py]: @Glutexo no, inlining is not at all what I meant. Take a look at the code suggestion.

File: app/xjoin.py (hunk 8 above)
Comment (Glutexo) [reply app/xjoin.py]: SQLAlchemy [doesn’t think](https://docs.sqlalchemy.org/en/13/core/engines.html?highlight=logging#configuring-logging) so. I tried to mimic its logging level behavior, so its consistent between the two data sources.

> sqlalchemy.engine - controls SQL echoing. set to logging.INFO for SQL query output, logging.DEBUG for query + result set output. These settings are equivalent to echo=True and echo="debug" on create_engine.echo, respectively.

Comment (jharting) [reply app/xjoin.py]: Looks like we also log the result on INFO level

File: api/host.py (hunk 4 above)
Comment (Glutexo) [reply api/host.py]: Oh, I see. The signature is not exactly same. Unfortunatly they differ in what exceptions they can trigger, which in my eyes is a part of the signature. I was thinking about refactoring that to introduce the exact change you suggest. I decided to keep that for some further refactoring cycle though.

If you insist on the refactoring, I can include it in this pull request. I wouldn’t want to do it the simplest way though: having two _except_ blocks for the same call would result in one of them always not being reachable. That seems like an antipattern to me.

File: app/serialization.py (hunk 7 above)
Comment (Glutexo) [reply app/serialization.py]: Actually, system profile facts [are not included](https://github.com/RedHatInsights/insights-host-inventory/blob/e81e7621fd0ce0f3e4452004eb41ba3685358ac6/swagger/api.spec.yaml#L682) in the _get_host_list_ output. Thus there is no need to even load them. Facts and tags should be there though, you’re right there.

File: api/host.py (hunk 4 above)
Comment (jharting) [reply api/host.py]: How do they differ in exceptions? From what I can see they both can raise `ValueError`. Not handling it for `get_host_list_xjoin` seems like an oversight not me.

File: app/xjoin.py (hunk 8 above)
Comment (Glutexo) [reply app/xjoin.py]: That should not be. Result should be DEBUG.

File: api/host.py (hunk 4 above)
Comment (Glutexo) [reply api/host.py]: 😯 You’re right. I’d swear I had a reason for that. This is worth adding a regression test.

**Comment by Glutexo** [comment]
Added facts.

**Comment by Glutexo** [comment]
Rebased.

File: test_api.py [hunk 12]
```diff
@@ -3103,246 +3112,712 @@ def test_tags_count_pagination(self):
...
+                        )
+
+        def test_response_invalid_pagination(self, is_enabled):
+            for page, per_page in [(0, 10), (-1, 10), (1, 0), (1, -5), (1, 101)]:
+                with self.subTest(page=page):
+                    with self.patch_with_empty_response():
+                        self.get(f"{TAGS_URL}?per_page={per_page}&page={page}", 400)
+
+    @patch("api.tag.is_enabled", return_value=True)
+    class TagsResponseTestCase(APIBaseTestCase):
```
Comment (jharting) [inline test_api.py]: why is this nested i.e. not top-level class?

File: test_api.py [hunk 13]
```diff
@@ -3103,246 +3112,712 @@ def test_tags_count_pagination(self):
...
-        self.get(f"{TAGS_URL}?per_page=2&page=3", 404)
+    @patch_with_response()
+    def test_response_pagination_index_error(self, graphql_query):
+        self.get(f"{HOST_URL}?per_page=2&page=3", 404)
+        graphql_query.assert_called_once()
 
-        graphql_query.assert_called_once_with(
-            TAGS_QUERY, {"order_by": "tag", "order_how": "ASC", "limit": 2, "offset": 4, "hostFilter": {"OR": ANY}}
+    @patch("api.tag.is_enabled", return_value=True)
+    class TagsRequestTestCase(XjoinRequestBaseTestCase):
```
Comment (jharting) [inline test_api.py]: why is this nested i.e. not top-level class?

File: test_api.py (hunk 12 above)
Comment (Glutexo) [reply test_api.py]: Argh! My humble guess is Black tried to guess its best when I forgot some closing bracket somewhere. Thanks for spotting! Fixed. 😇

Comment (jharting) [reply test_api.py]: @Glutexo don't forget to git push

**Pull request merged**